src/java.base/share/native/launcher/main.c

```diff
@@ -93,6 +93,64 @@ WinMain(HINSTANCE inst, HINSTANCE previnst, LPSTR cmdline, int cmdshow)
     __initenv = _environ;
 
 #else /* JAVAW */
+
+#include <sys/wait.h>
+
+static int is_checkpoint = 0;
+
+static void parse_checkpoint(const char *arg) {
+    if (!is_checkpoint) {
+        const char *checkpoint_arg = "-XX:CRaCCheckpointTo";
+        const int len = strlen(checkpoint_arg);
+        if (0 == strncmp(arg, checkpoint_arg, len)) {
+            is_checkpoint = 1;
+        }
+    }
+}
+
+static int wait_for_children() {
+    int status = 0;
+    pid_t pid;
+    do {
+        int st = 0;
+        pid = wait(&st);
+        if (!status && 0 < pid && WIFEXITED(st)) {
+            status = WEXITSTATUS(st);
```

Comment:
Some `WEXITSTATUS(st)` can end up 0, so the a next exited child can override this status.

I think here we should watch for the status of the g_child_pid, and ignoring statuses of other childs that are not our immediate java process child.

Reply:
I think Anton's right - we want to exit with the `g_child_pid`'s status so it's the only one we need to save away.  The rest of the wait() loop looks right in that it ensures we correctly wait on all children (even children of the g_child_pid) like an init process should.